File: metadata/timestamp_extraction.R
###
# Extracting timestamps from JPG files
# First and last timestamp in folder to get starttime and endtime of drone missions
###

library(jpeg)
library(exifr)

# Function to extract image timestamp using EXIF data

get_image_timestamp <- function(image_path) {
  exif_data <- exifr::exifr(image_path)
  
  if(!is.null(exif_data$DateTimeOriginal)) {
    return(exif_data$DateTimeOriginal)
  } else {
    return(NULL)
  }
}

# Function to process a folder and extract earliest and latest timestamps

process_folder <- function(folder_path) {
  earliest_timestamp <- NULL
  latest_timestamp <- NULL
  
  # Get a list of all files in the folder and its subfolders
  file_list <- list.files(path = folder_path, recursive = TRUE, full.names = TRUE)
  
  # Iterate through the list of files
  for (file in file_list) {
    # Check if the file is a JPEG image
    if (tolower(tools::file_ext(file)) == "jpg") {
      timestamp <- get_image_timestamp(file)
      
      # If the image has a valid timestamp in its EXIF data
      if (!is.null(timestamp)) {
        # Update earliest_timestamp if necessary
        if (is.null(earliest_timestamp) || timestamp < earliest_timestamp) {
          earliest_timestamp <- timestamp
        }
        
        # Update latest_timestamp if necessary
        if (is.null(latest_timestamp) || timestamp > latest_timestamp) {
          latest_timestamp <- timestamp
        }
      }
    }
  }
  
  return(list(earliest_timestamp = earliest_timestamp, latest_timestamp = latest_timestamp))
}


# Define the root folder containing the subfolder with JPEG files

root_folder <- "/lefodata/data/drone_missions/2021-09-02-sbl-cloutier-z3-P4RTK-WGS84"

# Process the root folder to get earliest and latest timestamps
timestamps <- process_folder(root_folder)

# Display the results
if (!is.null(timestamps$earliest_timestamp) && !is.null(timestamps$latest_timestamp)) {
  cat("Earliest timestamp:", timestamps$earliest_timestamp, "\n")
  cat("Latest timestamp:", timestamps$latest_timestamp, "\n")
} else {
  cat("No JPEG files with valid timestamps found.\n")
}

File: metadata/timestamp_extraction.py
import os
import pandas as pd
from PIL import Image 
from datetime import datetime, timedelta

def get_image_timestamp(image_path):
    try:
        img = Image.open(image_path)
        info = img._getexif()
        if info:
            timestamp = info.get(36867)  # EXIF tag for DateTimeOriginal
            if timestamp:
                return timestamp
    except Exception as e:
        print(f"Error processing {image_path}: {e}")
    return None

def process_folder(folder_path):
    earliest_timestamp = None
    latest_timestamp = None
    
    for root, _, files in os.walk(folder_path):
        for file in files:
            if file.lower().endswith(".jpg"):
                image_path = os.path.join(root, file)
                timestamp = get_image_timestamp(image_path)
                
                if timestamp:
                    # Parse the timestamp into a datetime object
                    try:
                        timestamp = datetime.strptime(timestamp, "%Y:%m:%d %H:%M:%S")
                    except Exception as e:
                        print(f"Error parsing timestamp {timestamp}: {e}")
                        continue  # Skip this file if timestamp parsing fails

                    # Apply 4-hour time offset
                    #timestamp += timedelta(hours=4)

                    if not earliest_timestamp or timestamp < earliest_timestamp:
                        earliest_timestamp = timestamp
                    if not latest_timestamp or timestamp > latest_timestamp:
                        latest_timestamp = timestamp

                    #print(f"Processed {image_path}, Timestamp:{timestamp}")
    
    return earliest_timestamp, latest_timestamp

if __name__ == "__main__":
    # Read the CSV file containing folder paths
    csv_file = "filename_mission_co2_2023.csv"
    df = pd.read_csv(csv_file)
    
    # Create a list to store results
    results = []
    
    # Process each folder path in the DataFrame
    for index, row in df.iterrows():
        folder_path = row["Folder Path"]
        print(f"Processing folder: {folder_path}")
        earliest, latest = process_folder(folder_path)
        
        if earliest and latest:
            results.append({
                "Folder Path": folder_path,
                "Earliest Timestamp": earliest,
                "Latest Timestamp": latest
            })
        else:
            print(f"No JPEG files with valid timestamps found in {folder_path}")
    
    # Create a DataFrame from the results
    results_df = pd.DataFrame(results)
    
    # Convert timestamps to RFC 3339 UTC format
    results_df["Earliest Timestamp"] = pd.to_datetime(results_df["Earliest Timestamp"]).dt.strftime('%Y-%m-%dT%H:%M:%SZ')
    results_df["Latest Timestamp"] = pd.to_datetime(results_df["Latest Timestamp"]).dt.strftime('%Y-%m-%dT%H:%M:%SZ')

    # Save the results DataFrame to an Excel file
    excel_output_file = "timestamps_rgb_co2_2023.xlsx"
    results_df.to_excel(excel_output_file, index=False)
    
    print("Timestamps saved to", excel_output_file)

#if __name__ == "__main__":
 #   root_folder = "F:/Dataset-2021-sbl/missions_drones/2021-09-02-sbl-cloutier-z3-P4RTK-WGS84"
  #  earliest, latest = process_folder(root_folder)
    
  #  if earliest and latest:
   #     print(f"Earliest timestamp: {earliest}")
    #    print(f"Latest timestamp: {latest}")
  #  else:
   #     print("No JPEG files with valid timestamps found.")

